library/std/src/ffi/c_str.rs

```diff
@@ -449,12 +449,18 @@ impl CString {
     ///     let c_string = CString::from_vec_unchecked(raw);
     /// }
     /// ```
+    #[inline(always)]
```

Comment:
What's the justification for this attribute? `inline(always)` is very rarely appropriate in libstd.

Reply:
My thinking was that the public method now simply forwards to the internal method, so it should be removed to match the performance of not having that extra layer of indirection.

Reply:
> the public method now simply forwards to the internal method

This is not accurate if `_from_vec_unchecked` already got inlined into `from_vec_unchecked`. In that situation `inline(always)` would be forcing a potentially much larger chunk of code than intended to be duplicated into every call site.

We tend to be skeptical of `inline(always)` without compelling benchmarks that demonstrate it's the only way to achieve certain commonly-necessary optimizations downstream.

Reply:
So inlining is bottom up rather than top down? I was thinking that `from_vec_unchecked` would get inlined first (so the call site now uses `_from_vec_unchecked`) and only *then* would the inliner consider `_from_vec_unchecked`. You're making it sound like `_from_vec_unchecked` could be inlined into `from_vec_unchecked` and then all of that gets inlined into the call site.

Reply:
Yeah it's bottom up: https://llvm.org/docs/Passes.html#inline-function-integration-inlining

> Bottom-up inlining of functions into callees.

Reply:
Dang, bummer. Fixed, except for `_from_bytes_with_nul_unchecked` which I changed to `always` since it's just a cast.